src/irmin-pack/pack_store.ml

```diff
@@ -127,92 +156,137 @@ module Maker
         Hashtbl.add roots (root, readonly) t;
         t
 
-    let v ?fresh ?readonly ?lru_size ~index root =
-      let t = unsafe_v ?fresh ?readonly ?lru_size ~index root in
+    let v ?fresh ?readonly ?lru_size ~index ~indexing_strategy root =
+      let t =
+        unsafe_v ?fresh ?readonly ?lru_size ~index ~indexing_strategy root
+      in
       Lwt.return t
 
-    let pp_hash = Irmin.Type.pp K.t
-    let decode_key = Irmin.Type.(unstage (decode_bin K.t))
-
     let io_read_and_decode_hash ~off t =
       let buf = Bytes.create K.hash_size in
       let n = IO.read t.pack.block ~off buf in
       assert (n = K.hash_size);
-      let _, v = decode_key (Bytes.unsafe_to_string buf) 0 in
-      v
+      let _, hash = decode_bin_hash (Bytes.unsafe_to_string buf) 0 in
+      hash
+
+    let pack_file_contains_key t k =
+      let offset = Key.to_offset k and length = Key.to_length k in
+      if
+        Int63.compare
+          (Int63.add offset (Int63.of_int length))
+          (IO.offset t.pack.block)
+        >= 0
+      then false
+      else
+        (* We read the hash explicitly as an integrity check: it's not
+           sufficient to assume that any offset within the file is valid. *)
+        let hash = io_read_and_decode_hash ~off:offset t in
+        let expected_hash = Key.to_hash k in
+        (* XXX: raise exception / log error in the [false] case here? *)
+        Hash.equal hash expected_hash
+
+    (* NOTE: may return false negatives, since the index may be partial. *)
+    (* let mem_via_hash t h =
+     *   Log.debug (fun l -> l "[pack] contains_hash %a" pp_hash h);
+     *   Tbl.mem t.staging h || Lru.mem t.lru h || Index.mem t.pack.index h *)
 
     let unsafe_mem t k =
-      [%log.debug "[pack] mem %a" pp_hash k];
-      Tbl.mem t.staging k || Lru.mem t.lru k || Index.mem t.pack.index k
+      [%log.debug "[pack] mem %a" pp_key k];
+      Tbl.mem t.staging (Key.to_hash k)
+      || Lru.mem t.lru (Key.to_hash k)
+      || pack_file_contains_key t k
 
     let mem t k =
       let b = unsafe_mem t k in
       Lwt.return b
 
-    let check_key k v =
-      let k' = Val.hash v in
-      if equal_key k k' then Ok () else Error (k, k')
+    let check_hash h v =
+      let h' = Val.hash v in
+      if equal_hash h h' then Ok () else Error (h, h')
 
-    exception Invalid_read
+    let _check_key k v = check_hash (Key.to_hash k) v
 
     let io_read_and_decode ~off ~len t =
-      if (not (IO.readonly t.pack.block)) && off > IO.offset t.pack.block then
-        raise Invalid_read;
       let buf = Bytes.create len in
       let n = IO.read t.pack.block ~off buf in
-      if n <> len then raise Invalid_read;
-      let hash off = io_read_and_decode_hash ~off t in
+      if n <> len then
+        invalid_read "Read %d bytes (at offset %a) but expected %d" n Int63.pp
+          off len;
+      let key_of_offset off =
+        let hash = io_read_and_decode_hash ~off t in
+        Pack_key.v ~hash ~offset:off ~length:len
```

Comment:
`~length:len` this uses the length of the parent inode for the child.

I arrived here because I'm having exactly the same problem for implementing an LRU for irmin 2.8: How to retrieve the length of the entry?

Reply:
Now resolved by attempting to read length headers from the pack file when possible (and otherwise waiting to check the index on `find`).